Creating a Hand Gesture Volume Controller using Python and Pycharm (#5403)
section-engineering-education/engineering-education: content/articles/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/index.md

- [Results](#results)
- [Conclusion](#Conclusion)

### Prerequisites
For you to follow through this article you need to:
- Be familiar with the Python programming language.
- Have **Pycharm** installed on their computer. If this is not the case, download it from here. [Pycharm Community Edition ](https://www.jetbrains.com/pycharm/download/)

### Python libraries we will need
The libraries that we will be interacting with while creating our hand gesture volume controller will be the following:
- *numpy* This library will help us work with arrays.
- *cv2* We will use this library to capture an image using the webcam and convert the image to RGB.
- *mediapipe* This is an open-source library developed by Google. It is used for both face and gesture recognition. For the matter of this tutorial, we will use it for hand gesture recognition.
- *math* We will use this library to find the distance between point number 4 (the thumb) and point number 8 (the index finger) using hypotenuse.
- *ctypes* and *comtypes* These two libraries are depended upon by pycaw. Ctypes to provide C language compatible data types. Comtypes bases on the ctypes FFI(Foreign Function Interface) library.

> - *pycaw* We will need this library to access the device's speaker and its master volume.

Reviewer: Instead of writing this section separately with the code, I'd rather you explain each library after its `pip install` equivalent.

# Detecting, initializing and configuring the hands.
mpHands = mp.solutions.hands
hands = mpHands.Hands()
mpDraw = mp.solutions.drawing_utils
# Accessing the speaker through pycaw.
devices = AudioUtilities.GetSpeakers()
interface = devices.Activate(IAudioEndpointVolume._iid_, CLSCTX_ALL, None)
volume = cast(interface, POINTER(IAudioEndpointVolume))

volMin, volMax = volume.GetVolumeRange()[:2]

while True:
    success, img = cap.read()  # if the camera we have specified works, we will capture an image.
    imgRGB = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)  # converting the image the camera captures to RGB Image.
    results = hands.process(imgRGB)  # completes the processing of the converted image.

    lmList = []  # creating an empty list.
    if results.multi_hand_landmarks:  # A list of all the detected hands.
        #  Creating a loop so we can the list above so as to get the corresponding flag bit of each hand.
        for handlandmark in results.multi_hand_landmarks:
            for id, lm in enumerate(handlandmark.landmark):  # We are adding a counter and then returning it.
                # Finding the finger joints of the hand

>                 h, w, _ = img.shape

Reviewer: @leeseshawn Instead of pasting the whole code and writing the description using the comments, you can separate the code into small sections and use the comments to write complete sentence/paragraph describing what each section does. 

Please refer to this [tutorial](https://github.com/section-engineering-education/engineering-education/pull/5163/files) for reference

        cv2.line(img, (x1, y1), (x2, y2), (255, 0, 0), 3)  # Drawing a line between the index finger and the thumb.

        length = hypot(x2 - x1, y2 - y1)  # Finding distance between the tips of the thumb and index finger using
                                            # hypotenuse.
        vol = np.interp(length, [15, 220], [volMin, volMax])  # Hand range 15 - 220
        print(vol, length)
        volume.SetMasterVolumeLevel(vol, None)  # Setting the master volume in accordance to hand range.
                                                # Volume range -63.5 - 0.0

    cv2.imshow('Image', img)  # Showing the real time video which shows how the user is adjusting the volume using their
                                # hand gesture
    if cv2.waitKey(1) & 0xff == ord('q'):  # Terminate the program if the user does press letter q.
        break
```
### Results
When the code above has run to completion without any errors, the results will be as those shown in the screenshot below.

![Results](/engineering-education/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/results.png)

> 

Reviewer: You could provide a gif image demonstrating these results instead. Providing only a screenshot doesn't suffice.

We need to install the libraries we need to use in our project.

`numpy` will help us work with arrays. To install it, open the terminal and run the following command: 

```bash
pip install numpy
```
Repeat the same process for the other libraries.

```bash
pip install opencv-python
```
This library is imported as `cv2`. We will use it to capture an image using the webcam and convert the image to `RGB`.

```bash
pip install mediapipe
```
This is an open-source library developed by Google. It's used for both face and gesture recognition. For this tutorial, we'll use it for hand gesture recognition.

```bash
pip install pycaw
```
We'll need this library to access the device's speaker and its master volume.

```bash
pip install python-math
```
We'll use this library to find the distance between point number `4` (the thumb) and point number `8` (the index finger) using hypotenuse.

```bash
pip install gpib-ctypes, comtypes
```
`pycaw` depends on these two libraries. `Ctypes` providea `C` language compatible data types. `Comtypes` bases on the `ctypes` **FFI**(Foreign Function Interface) library.

Now let us jump into coding. This is where the most fun is as we get hands-on.

> ```python

Reviewer 2: You haven't specified in which file to place the code

import cv2
import mediapipe as mp
from math import hypot
from ctypes import cast, POINTER
from comtypes import CLSCTX_ALL
from pycaw.pycaw import AudioUtilities, IAudioEndpointVolume
import numpy as np
```
In the above code segment, we are importing each library we did discuss into our project.

```python
cap = cv2.VideoCapture(0) 
```
We are getting the video input from our computer's primary camera. If you are using any other camera, replace the number `0` with that of the camera you are using.

### Detecting, initializing, and configuring the hands
```python
mpHands = mp.solutions.hands
hands = mpHands.Hands()
mpDraw = mp.solutions.drawing_utils
```
- `mp.solutions.hands`: We are calling on the `mediapipe` hand module to detect the hands from the video input we did get.

> - `mpHands.Hands()`: This completes the initialization and configuration of the hands we did detect.

Reviewer 2: I am seeing several sentences like this which are somehow vague and grammatically incorrect. Please check on them. For instance, instead of `...we did detect` you can say `...we detected`. Some statements are in passive voices. Change them to active voice.

Also, check your dictum. Make use of simple words that best describe what you want to get across.

<!--more-->
At this point, a hand gesture volume controller would come in handy. You don't have to be concerned about the state of your hands. You just have to use your fingers and adjust the computer's volume to your level of satisfaction.

### Table of contents
- [Prerequisites](#prerequisites)
- [Referral hand image](#referral-hand-image)
- [Creating a hand gesture volume controller](#creating-a-hand-gesture-volume-controller)
- [Results](#results)
- [Conclusion](#Conclusion)

### Prerequisites
For you to follow through this article you need to:
- Be familiar with the Python programming language.
- Have **Pycharm** installed on their computer. If this is not the case, download it from here. [Pycharm Community Edition ](https://www.jetbrains.com/pycharm/download/)

### Introduction
**Python** has libraries that will help us create our hand gesture volume controller. **Pycharm** is the IDE that we will use to run our python code. We will be using the **Pycharm** community version IDE. It is free to download and use from the internet. 

> In this tutorial, we will discuss the use of the python libraries we will need, how to use python to track hand gestures, and finally how to create a hand gesture volume controller using Python and Pycharm. A person using Windows, Linux, or Mac OS can follow through.

Reviewer: ```suggestion
This tutorial will discuss the use of the python libraries we will need, how to use python to track hand gestures, and finally how to create a hand gesture volume controller using Python and Pycharm. You can follow along using either a Windows, Linux, or Mac OS machine.
```

### Referral hand image
![hand](/engineering-education/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/hand.jpg)

> *[Image Source: Aniketiq](https://www.bing.com/images/search?view=detailV2&ccid=A16aeeUy&id=2B00C1DE13C6EFD132E8FCABF5CF0A11695E1FCA&thid=OIP.A16aeeUyQVRHqGt_rphzJQHaCl&mediaurl=https%3a%2f%2fcdn.hashnode.com%2fres%2fhashnode%2fimage%2fupload%2fv1629398588657%2fboG68w85n.png%3fauto%3dcompress%2cformat%26format%3dwebp&cdnurl=https%3a%2f%2fth.bing.com%2fth%2fid%2fR.035e9a79e532415447a86b7fae987325%3frik%3dyh9eaREKz%252fWr%252fA%26pid%3dImgRaw%26r%3d0&exph=538&expw=1543&q=mediapipe+hand+tracking+points&simid=607990468186358942&FORM=IRPRST&ck=C941CC346294E5FC99172E9ED31C9E5B&selectedIndex=4)*

Reviewer: ```suggestion
*[Image Source: MediaPipe](https://google.github.io/mediapipe/solutions/hands.html)*
```

Reviewer: That link is too long. Consider replacing it with this link instead.

```
*numpy* will help us work with arrays.

![Terminal](/engineering-education/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/terminal.png)

The above screenshot shows where to type the command for numpy installation.

Wait for a few seconds till numpy is installed successfully. We are going to repeat the same steps for the other libraries. We will only replace the command part. Use the following commands for the respective libraries.
```bash
 pip install opencv-python

```
 This library is imported as *cv2*. We will use this library to capture an image using the webcam and convert the image to RGB.
 ```bash
pip install mediapipe

 ```
 This is an open-source library developed by Google. It is used for both face and gesture recognition. For the matter of this tutorial, we will use it for hand gesture recognition.
 ```bash
pip install pycaw
 ```
  We will need this library to access the device's speaker and its master volume.
 ```bash
pip install python-math
 ```
 We will use this library to find the distance between point number 4 (the thumb) and point number 8 (the index finger) using hypotenuse.
 ```bash
pip install gpib-ctypes, comtypes
 ```

>  These two libraries are depended upon by pycaw. Ctypes to provide C language compatible data types. Comtypes bases on the ctypes FFI(Foreign Function Interface) library.

Reviewer: Consider using (``) when referring to libraries. For example, `pycaw`. It helps in highlighting them.

Now let us jump into coding. This is where the most fun is as we get hands-on.

```python
import cv2
import mediapipe as mp
from math import hypot
from ctypes import cast, POINTER
from comtypes import CLSCTX_ALL
from pycaw.pycaw import AudioUtilities, IAudioEndpointVolume
import numpy as np
```
In the above code segment, we are importing each library we did discuss into our project.

```python
cap = cv2.VideoCapture(0) 
```

> We are getting the video input from our computer's primary camera. If you are using any other camera, replace the number 0 with that of the camera you are using.

Reviewer: Also, use (``) when highlighting numbers. For example, `0`

### Detecting, initializing, and configuring the hands
```python
mpHands = mp.solutions.hands
hands = mpHands.Hands()
mpDraw = mp.solutions.drawing_utils
```
- `mp.solutions.hands` : 
We are calling on mediapipe hand module to detect the hands from the video input we did get.
- `mpHands.Hands()` :
This completes the initialization and configuration of the hands we did detect.
- `mp.solutions.drawing_utils` : 
Draws the connections and landmarks of the hand image. 

### Accessing the speaker through pycaw
```python
devices = AudioUtilities.GetSpeakers()
interface = devices.Activate(IAudioEndpointVolume._iid_, CLSCTX_ALL, None)
volume = cast(interface, POINTER(IAudioEndpointVolume))
```

> This are the initializations we need for pycaw. The library developer provides them so we will not change anything. They are found [here](https://github.com/AndreMiras/pycaw).

Reviewer: Please use `These` not `this`. For example, "This is.....these are"

### Finding the volume range between the minimum and maximum volume
```python
volMin, volMax = volume.GetVolumeRange()[:2]
```
We are calling the pycaw `volume.GetVolumeRange()` function to find the volume range between the minimum and maximum volume.
### Capturing an image from our camera and converting it to an RGB image
```python
while True:
    success, img = cap.read()
    imgRGB = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    results = hands.process(imgRGB)
```

> - `success, img = cap.read()`: Checking whether the camera we have specified works, we will capture an image.
> - `imgRGB = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)`: Converting the image the camera captures to RGB image.

Reviewer: ```suggestion
- `success, img = cap.read()`: Checks whether the camera we have specified works. If it works,  we will capture an image.
- `imgRGB = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)`: Converts the image captured by the camera to `RGB` image.
```

```
In the above code, we are using the cv2.line function to draw a line between point four of the hand, and point 8.
- `img`:
We are passing our image to the function.
- `(x1, y1), (x2, y2)`:
We are specifying that the line we will draw will connect point 4 (`(x1, y1)`) which is the tip of the thumb and point 8 (`(x2, y2)`) which is the tip of the index finger.
- `(255, 0, 0)`:
This is the color of the line.
- `3`:
This is the thickness of the line.

### Finding the distance between points 4 and 8
```python
        length = hypot(x2 - x1, y2 - y1)
```
In the above code, we are finding the distance between the tip of the thumb and the index finger using hypotenuse. We achieve this by calling the math hypot function and passing the difference between x2 and x1 and the difference between y2 and y1.
### Converting the hand range to the volume range
```python                                          
        vol = np.interp(length, [15, 220], [volMin, volMax])
        print(vol, length)
```
We are calling on the numpy function `np.interp` to convert the hand range to the volume range. The arguments used are:
- `length`: This is the value we want to convert.

> - `[150 - 220]`: This is the hand range.

Reviewer: Is the range [15, 220] or [150, 220]? That bit is not clear

- `[volMin, volMax]`: Giving the range to which we want to convert.

### Setting the master volume
```python        
        volume.SetMasterVolumeLevel(vol, None)  
```
We are setting the master volume level in accordance with the hand range. We achieve this by passing `vol` which is the value of the hand range we did convert to volume range.

### Displaying the video output used to interact with the user
```python
    cv2.imshow('Image', img) 
```
The above code shows the real-time video showing the user interacting with the program. i.e The user uses the thumb finger and the index finger to control the volume.

### Terminating the program
```python
    if cv2.waitKey(1) & 0xff == ord('q'): 
        break
```
The above code will terminate the program if the user presses the q key.
### Results
When the code above has run to completion without any errors, the results will be as those shown in the screenshot below.

![Results](/engineering-education/creating-a-hand-gesture-volume-controller-using-python-and-pycharm/results.gif)

### Conclusion 

> You now have all the skills required to create a hand gesture volume controller. Do your work while listening to your favorite music. You don't have to worry about dirtying your hands. By just a gesture of your hand, you can control the volume level of your music. You

Reviewer: Remove the `You` at the end of the sentence. You need to proofread your work to avoid unnecessary corrections.

### Finding the volume range between the minimum and maximum volume
```python
volMin, volMax = volume.GetVolumeRange()[:2]
```
We are calling the `pycaw` `volume.GetVolumeRange()` function to find the volume range between the minimum and maximum volume.

### Capturing an image from our camera and converting it to an RGB image
```python
while True:
    success, img = cap.read()
    imgRGB = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    results = hands.process(imgRGB)
```
`success, img = cap.read()`: checks whether the camera we have specified works. If it works, we will capture an image.

`imgRGB = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)`: converts the image captured by the camera to an `RGB` image.

` results = hands.process(imgRGB)`: completes the processing of the image that was converted to `RGB`.

### Checking whether we have multiple hands in our input
```python
lmList = []
if results.multi_hand_landmarks: 
```
`lmList = []`: this code creates an *empty list* which will store the list of elements of the hands detected by the `mediapipe` hand module, i.e., the number of points on the hand.

`if results.multi_hand_landmarks:` checks whether the input has multiple hands.

> ### Creating a for loop to manipulate each hand

Reviewer 2: There is an issue with how you are explaining your code. In the prerequisites, you specified that knowledge in python is required. So I don't find it necessary to repeat each and every line of code and what it does. Since the readers know about syntax and control structures, you can give an overview of what the code does and what the output may be in terms of short paragraphs. The same way you have done in the `Finding the distance between points 4 and 8` and `Terminating the program` sections. 

Again, go through published tutorials to see how they go about article writing.

### Specifying the points of the thumb and middle finger we will use
```python
if lmList != []:
    x1, y1 = lmList[4][1], lmList[4][2]
    x2, y2 = lmList[8][1], lmList[8][2]
```
`if lmlist != []`: we use this statement to specify that the number of elements should not be null.
 
`x1, y1 = lmList[4][1], lmList[4][2]` : we are assigning variables `x1` and `y1` the `x` and `y` coordinates of point `4` respectively. This is the tip of the thumb. 
 
`x2, y2 = lmList[8][1], lmList[8][2]`: we are assigning variables `x2` and `y2` the `x` and `y` coordinates of point `8` respectively. This is the tip of the index finger. 
 
Refer to the hand image diagram we discussed to identify the [points](#referral-hand-image).

### Drawing a line between the tip of the thumb and the tip of the index finger
```python                                            
       cv2.circle(img, (x1, y1), 15, (255, 0, 0), cv2.FILLED)  
       cv2.circle(img, (x2, y2), 15, (255, 0, 0), cv2.FILLED)  
```
`cv2.circle(img, (x1, y1), 4, (255, 0, 0), cv2.FILLED)`: we are calling the `cv2.circle` and passing our image `img`.

` (x1, y1)`: it specifies that we will draw the circle at the tip of the thumb.

> `15`: this is the *radius* of the circle.

Reviewer 2: As you can see, this section isn't really aesthetically pleasing to read as there are way too short statements. You can just give your explanations in paragraphs without repeating the code blocks.
